### PR TITLE
[WIP]: Marathon Store take 2

### PR DESCRIPTION
**Pull request by JayH5**
Make the Store (previously, State) a standalone GenServer that stores apps and tasks and forwards events when something changes.

### CODE REVIEW

File: lib/relay/marathon/store.ex [hunk 1]
```diff
@@ -1,96 +1,191 @@
...
+  def start_link(opts \\ []) do
+    GenServer.start_link(__MODULE__, :ok, opts)
+  end
+
+  @doc """
+  Update an app in the Store. The app is only added if its version is newer than
+  any existing app.
+  """
+  @spec update_app(pid, App.t) :: :ok
+  def update_app(server, %App{} = app), do: GenServer.call(server, {:update_app, app})
```
[the remark this answers is not in the document]
Comment (JayH5) [reply lib/relay/marathon/store.ex]: It's not really relevant to the thing calling `update`.. we could return the existing (or non-existing) app. I'm also not sure which atoms are typical return values other than `:ok` and `:error`.

File: lib/relay/marathon/store.ex [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+      |> Enum.sort(fn %App{id: id1}, %App{id: id2} -> id1 < id2 end)
+    end
+
+    @spec get_apps_and_tasks(t) :: [{App.t(), [Task.t()]}]
+    def get_apps_and_tasks(%__MODULE__{} = state) do
+      state
+      |> get_apps()
+      |> Enum.map(fn %App{id: app_id} = app ->
+        tasks =
+          state.app_tasks[app_id]
```
Comment (jerith) [inline lib/relay/marathon/store.ex]: Would it make sense to split this `fn` out into a separate function?

File: lib/relay/marathon/store.ex [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+    {:ok, {resources, State.new()}}
+  end
+
+  def handle_call({:update_app, %App{id: id, version: version} = app}, _from, {resources, state}) do
+    {old_app, new_state} = State.get_and_update_app(state, app)
+
+    case old_app do
+      %App{version: existing_version} when version > existing_version ->
+        Log.debug("App '#{id}' updated: #{existing_version} -> #{version}")
+        notify_updated_app(resources, new_state)
```
Comment (jerith) [inline lib/relay/marathon/store.ex]: This looks like duplicated logic. Maybe we need an `update_app` that replies with the new state and `:new | :updated | :unchanged` or something? I guess we'd need to return the existing version as well, unless we put the logging in the state management code.

Alternatively, maybe we should move the guts of `State.get_and_update_app` here and make the functions it calls public in the `State` module?

File: lib/relay/marathon/store.ex [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+
+          nil ->
+            Log.info("Task '#{id}' with version #{version} added")
+            notify_updated_task(resources, new_state)
+        end
+
+        new_state
+      rescue
+        KeyError ->
+          Log.warn("Unable to find app '#{app_id}' for task '#{id}'. Task update ignored.")
```
Comment (jerith) [inline lib/relay/marathon/store.ex]: Is this `try`/`rescue` really necessary? I'd prefer to either structure the update code in a way that makes this an expected case or decide that it's a "can't happen" case and crash if we see it.

File: lib/relay/marathon/store.ex (hunk 3 above)
Comment (JayH5) [reply lib/relay/marathon/store.ex]: Yep it is duplicated. I was trying to separate the ideas of "the `State` keeps itself consistent" from "the `GenServer` logs what has happened".

File: lib/relay/marathon/store.ex (hunk 4 above)
Comment (JayH5) [reply lib/relay/marathon/store.ex]: In the older version of the store (when it was just called state), this used to be a `!` function. It's true that this shouldn't happen: a task should only be considered at all if we have an app for it stored. But I guess I wasn't sure how to handle this in the case of the thing being a `GenServer`.

Perhaps whatever is feeding apps and tasks to the store should just send it (update/delete) every task it receives in an event and the store should just ignore tasks for which it doesn't have the corresponding app. The thing sending the tasks doesn't have any way to determine whether a task is relevant without knowing about the app for the task.

File: lib/relay/marathon/store.ex [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+          %Task{version: version} ->
+            Log.info("Task '#{id}' with version #{version} deleted")
+            notify_updated_task(resources, new_state)
+
+          nil ->
+            Log.debug("Task '#{id}' not present/already deleted")
+        end
+
+        new_state
+      rescue
```
Comment (jerith) [inline lib/relay/marathon/store.ex]: I don't really like this `try`/`rescue` here -- it indicates that the `State` API doesn't really match what we're trying to use it for.

File: lib/relay/marathon/store.ex [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+      do: map |> Map.to_list() |> Enum.sort() |> Enum.map(&elem(&1, 1))
+
+    @spec get_app(t, String.t()) :: App.t() | nil
+    def get_app(%__MODULE__{apps: apps}, app_id), do: Map.get(apps, app_id)
+
+    @spec get_and_update_app(t, App.t()) :: {App.t() | nil, t}
+    def get_and_update_app(%__MODULE__{apps: apps} = state, %App{id: id, version: version} = app) do
+      case Map.get(apps, id) do
+        # App is newer than existing app, update the app
+        %App{version: existing_version} = existing_app when version > existing_version ->
```
[the remark this answers is not in the document]
Comment (JayH5) [reply lib/relay/marathon/store.ex]: The answer is no. :(

**Pull request merged**